litgpt/generate/base.py

```diff
@@ -194,6 +194,120 @@ def generate_fn(
 
 
+# TODO: Make include_eos work.
+# TODO: Rewrite unbatched generate_fn to use batched_generate_fn.
+@torch.inference_mode()
+def batched_generate_fn(
+    model: GPT,
+    prompts: torch.Tensor,
+    max_returned_tokens: int,
+    *,
+    sample_args: Union[list[dict], dict],
+    stop_tokens: Tuple[List[int], ...] = (),
+    include_prompt: bool,
+    include_eos: bool,
+) -> Iterator[list[Union[torch.Tensor, None]]]:
+
+    if isinstance(sample_args, dict):
+        sample_args = [sample_args] * len(prompts)
+
+    if prompts.ndim == 1:
+        prompts = prompts.unsqueeze(0)
+
+    assert prompts.ndim == 2, "Prompts must be a 2D tensor."
+    batch_size = prompts.size(0)
+    max_prompt_size = prompts.size(1)
+    device = prompts.device
+
+    assert all(prompt.size(0) == max_prompt_size for prompt in prompts), "For now, prompts must have the same length."
+    assert all(prompt.device == prompts[0].device for prompt in prompts), "Prompts must be on the same device."
+
+    assert max_returned_tokens > max_prompt_size, f"Not enough space for {max_prompt_size} prompt tokens in a context length of {max_returned_tokens}."
+    if model.max_seq_length < max_returned_tokens - 1:
+        raise NotImplementedError(f"max_seq_length {model.max_seq_length} needs to be >= {max_returned_tokens - 1}")
+
+    # Yield the prompts if include_prompt is True
+    if include_prompt:
+        # TODO: Prompt length is padded, but they shouldn't all be the same length.
+        for i in range(max_prompt_size):
+            yield [prompt[i].view(-1) for prompt in prompts]
+
+    stop_progresses = [[0] * len(stop_tokens) for _ in range(batch_size)] # [batch_size, ~len(stop_tokens)]
+    stop_idxes = [-1] * batch_size
+    yielded_idx = 0
+
+    # Generate output tokens.
+    # The first token generated is the prefill token.
+    # The input_pos for this token is the width of the entire prompt.
+    # For subsequent iterations, it's the index in the context for the token that we're generating.
+    token_lists = [[] for _ in range(batch_size)]
+    tokens: torch.Tensor = prompts
+    prefill_token = True
+    input_pos = torch.arange(0, max_prompt_size, device=device, dtype=torch.int64)
+    for current_idx in range(max_returned_tokens - max_prompt_size):
+
+        # Generate the next token for each prompt in the batch.
+        # This is of shape [batch_size, 1].
+        tokens = batched_next_token(model, input_pos, tokens, sample_args)
+        for i in range(batch_size):
+            token_lists[i].append(tokens[i])
+        int_tokens = [token.item() for token in tokens]
+
+        # Check for stop sequences
+        # For each stop sequence, we keep a running total of how many are matched in stop_progress.
+        # If the current token matches the next token in the stop sequence, we increment the
+        # running total and hold off on yielding the token.
+        for batch_idx, int_token in enumerate(int_tokens):
```

Comment:
That we're basically doing this per batch-index makes me wonder what was the advantage of switching it to this loop from the one-line any condition.